src/microbe_stage/CompoundCloudPlane.cs

```diff
@@ -347,380 +425,168 @@ public void ConvertToCloudLocal(Vector3 worldPosition, out int x, out int y)
     {
-        var xComponent = this;
-        if (x < 0)
-        {
-            xComponent = LeftCloud;
-        }
-        else if (x >= Size)
-        {
-            xComponent = RightCloud;
-        }
+        float a = delta * Constants.CLOUD_DIFFUSION_RATE;
 
-        if (xComponent == null)
-            return;
-
-        var yComponent = xComponent;
-        if (y < 0)
-        {
-            yComponent = xComponent.UpperCloud;
-        }
-        else if (y >= Size)
+        for (int x = x0; x < x0 + width; x++)
         {
-            yComponent = xComponent.LowerCloud;
+            for (int y = y0; y < y0 + height; y++)
+            {
+                OldDensity[x, y] =
+                    Density[x, y] * (1 - a) +
+                    (Density[x, y - 1] + Density[x, y + 1] + Density[x - 1, y] + Density[x + 1, y]) * (a / 4);
+            }
         }
-
-        if (yComponent == null)
-            return;
-
-        x = (x + Size) % Size;
-        y = (y + Size) % Size;
-
-        yComponent.slots[slotIndex].Density[x, y] += value;
     }
 
-    private class Slot
+    private void PartialDiffuseEdges(int x0, int y0, int width, int height, float delta)
     {
-        public float[,] Density;
-        public float[,] OldDensity;
-        public Compound Compound;
-        private readonly int size;
-        private readonly int resolution;
-        private readonly FluidSystem fluidSystem;
-        private readonly CompoundCloudPlane parentPlane;
-        private readonly int index;
-
-        public Slot(Compound compound, int size, int resolution,
-            FluidSystem fluidSystem, CompoundCloudPlane parent, int index)
-        {
-            this.size = size;
-            this.resolution = resolution;
-            this.fluidSystem = fluidSystem;
-            Compound = compound;
-            parentPlane = parent;
-            this.index = index;
-
-            if (size <= 0)
-                return;
-
-            // For simplicity all the densities always exist, even on the unused clouds
-            Density = new float[size, size];
-
-            // Except the old density can be easily ignored so this saves some memory
-            if (Compound != null)
-                OldDensity = new float[size, size];
-        }
+        float a = delta * Constants.CLOUD_DIFFUSION_RATE;
 
-        public void Clear()
+        for (int x = x0; x < x0 + width; x++)
         {
-            for (int x = 0; x < size; ++x)
+            for (int y = y0; y < y0 + height; y++)
             {
-                for (int y = 0; y < size; ++y)
-                {
-                    Density[x, y] = 0.0f;
-                }
+                OldDensity[x, y] =
+                    Density[x, y] * (1 - a) +
+                    (Density[x, (y - 1 + Size) % Size] +
+                    Density[x, (y + 1) % Size] +
+                    Density[(x - 1 + Size) % Size, y] +
+                    Density[(x + 1) % Size, y]) * (a / 4);
```

Comment:
There is a ton of magic numbers used in the math in this file. Could at least some of these be changed to be named constants? Like here is the 4, the same as the clouds in one constant? Or is it an unrelated value that just happens to be 4? Using constants would make the code much more understandable.

Reply:
I changed, and it looks horrible lol.
It seems to have increased my FPS for some reason, tho ~here's hoping that doesn't mean something broke along the way~